### PR TITLE
Issue #3466: Clear cache in DetailAST before changing its structure 

### PR DESCRIPTION
**Pull request by MEZk**
#3466

### CODE REVIEW

**Comment by rnveach** [comment]
@MEZk 
what about moving the reset of `childCount` into `clearLazyLoadCache`?
We clear it on `setFirstChild` but not on `addChild` which it seems like we should do.

**Comment by MEZk** [comment]
@rnveach 
Yes, you are right. But for now I decided to concentrate on 'branchContains' method and implement UT for the method. And as you can see, UT is failing and I still do not understand why.

**Comment by rnveach** [comment]
@MEZk 
`testClearLazyLoadCacheInDetailAST` looks like it is failing in `CheckWhichRequiresCommentNodes.visitToken` because your missing the fact that the comment isn't part of `METHOD_DEF` but is in `MODIFIERS`.
AST Print:
````
    |--METHOD_DEF -> METHOD_DEF [8:4]
    |  |--MODIFIERS -> MODIFIERS [8:4]
    |  |  |--BLOCK_COMMENT_BEGIN -> /* [5:4]
    |  |  |  |--COMMENT_CONTENT -> *\n    * Javadoc comment\n      [5:6]
    |  |  |  `--BLOCK_COMMENT_END -> */ [7:5]
    |  |  `--LITERAL_PUBLIC -> public [8:4]
````

if you change your if statements to something like `ast.getFirstChild().branchContains` it passes then.

**Comment by MEZk** [comment]
@rnveach 
It is not a solution. 'branchTokenTypes' in DetailAST will be initialized for AST without comment nodes and than will be used for AST with comment nodes. That is a problem, I guess.

Yes, UT will not fail, but If I comment out the following lines:

https://github.com/MEZk/checkstyle/blob/2cc394e7c1917e1c3be8bf55953e5947cf4ecfc3/src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java#L685-L686
and
https://github.com/MEZk/checkstyle/blob/2cc394e7c1917e1c3be8bf55953e5947cf4ecfc3/src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java#L691-L691
the test will also pass. 

Could you take a closer look at these code fragments?
https://github.com/MEZk/checkstyle/blob/fbf3f95c2273be23b966cba6e6ef89390a6fb020/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java#L174-L180
and
https://github.com/MEZk/checkstyle/blob/2cc394e7c1917e1c3be8bf55953e5947cf4ecfc3/src/main/java/com/puppycrawl/tools/checkstyle/api/DetailAST.java#L333-L333

**Comment by MEZk** [comment]
@romani 
1) Added UT.
2) Fixed the problem. 

Solution:
Lazy-load cache should be cleared for all parents of the current DetailAST object in all methods which change the structure of the tree.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/DetailAST.java [hunk 1]
```diff
@@ -395,4 +401,14 @@ public DetailAST getFirstChild() {
         return (DetailAST) super.getFirstChild();
     }
 
+    /**
+     * Clears lazy-load cache for all parents of the current {@link DetailAST} instance.
+     */
+    private void clearLazyLoadCache() {
+        DetailAST prevParent = getParent();
+        while (prevParent != null) {
+            prevParent.branchTokenTypes = null;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/DetailAST.java]: `clearLazyLoadCache`  vs `branchTokenTypes = null;`

it is not the only cache that we have in that class , so please be specific in the name - `clearBranchTokenTypesCache`

**Comment by MEZk** [comment]
@romani 
>t is not the only cache that we have in that class , so please be specific in the name - clearBranchTokenTypesCache

Renamed.
Please, approve the issue.

**Pull request merged**